verifier/providerverifier.go

```diff
@@ -0,0 +1,305 @@
+	"github.com/openpubkey/openpubkey/gq"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+	oidcclient "github.com/zitadel/oidc/v2/pkg/client"
+)
+
+type DefaultProviderVerifier struct {
+	issuer          string
+	commitmentClaim string
+	options         ProviderVerifierOpts
+}
+
+type ProviderVerifierOpts struct {
+	// If ClientID is specified, then verification will require that the ClientID
+	// be present in the audience ("aud") claim of the PK token payload
+	ClientID string
+	// Specifies whether to skip the Client ID check, defaults to false
+	SkipClientIDCheck bool
+	// Custom function for discovering public key of Provider
+	DiscoverPublicKey func(ctx context.Context, kid string, issuer string) (JSONWebKey, error)
+	// Allows for successful verification of expired tokens
+	SkipExpirationCheck bool
+}
+
+// Creates a new ProviderVerifier with required fields
+//
+// issuer: Is the OpenID provider issuer as seen in ID token e.g. "https://accounts.google.com"
+// commitmentClaim: the ID token payload claim name where the cicHash was stored during issuance
+func NewProviderVerifier(issuer, commitmentClaim string, options ProviderVerifierOpts) *DefaultProviderVerifier {
+	v := &DefaultProviderVerifier{
+		issuer:          issuer,
+		commitmentClaim: commitmentClaim,
+		options:         options,
+	}
+
+	// If no custom DiscoverPublicKey function is set, set default
+	if v.options.DiscoverPublicKey == nil {
+		v.options.DiscoverPublicKey = DiscoverProviderPublicKey
+	}
+
+	return v
+}
+
+func (v *DefaultProviderVerifier) Issuer() string {
+	return v.issuer
+}
+
+func (v *DefaultProviderVerifier) VerifyProvider(ctx context.Context, pkt *pktoken.PKToken) error {
+	// Check whether Audience claim matches provided Client ID
+	// No error is thrown if option is set to skip client ID check
+	if err := verifyAudience(pkt, v.options.ClientID); err != nil && !v.options.SkipClientIDCheck {
+		return err
+	}
+
+	alg, ok := pkt.ProviderAlgorithm()
+	if !ok {
+		return fmt.Errorf("provider algorithm type missing")
+	}
+
+	switch alg {
+	case gq.GQ256:
+		if err := VerifyGQSig(ctx, pkt); err != nil {
+			return fmt.Errorf("error verifying OP GQ signature on PK Token: %w", err)
+		}
+	case jwa.RS256:
+		opToken, err := pkt.Compact(pkt.Op)
+		if err != nil {
+			return err
+		}
+
+		pubKey, err := v.ProviderPublicKey(ctx, opToken)
+		if err != nil {
+			return fmt.Errorf("failed to get OP public key: %w", err)
+		}
+
+		token, err := pkt.Compact(pkt.Op)
+		if err != nil {
+			return err
+		}
+
+		if _, err := jws.Verify(token, jws.WithKey(alg, pubKey)); err != nil {
+			return err
+		}
+	}
+
+	if err := v.verifyCommitment(pkt); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// This function takes in an OIDC Provider created ID token or GQ-signed modification of one and returns
+// the associated public key
+func (v *DefaultProviderVerifier) ProviderPublicKey(ctx context.Context, token []byte) (JSONWebKey, error) {
```

Comment:
This PR changes `func DiscoverPublicKey(ctx context.Context, headers jws.Headers, issuer string) (crypto.PublicKey, error)`  to  `func (v *DefaultProviderVerifier) ProviderPublicKey(ctx context.Context, token []byte) (JSONWebKey, error)`

This change breaks the future ability to look up a PublicKey based on a `jtk` public header claim as the new ProviderPublicKey function takes a compact Token. Compact Tokens exclude public header claims. The `jtk` public header functionality was added specifically to enable the ability to look up an OP Public Key when the OP JWKS has no kid (see https://github.com/openpubkey/openpubkey/pull/73 )

This is also used by a our GQ Signature header format:
https://github.com/openpubkey/openpubkey/pull/66/files

In the current codebase, we throw an error if the kid does not exist rather than falling back to the jtk claim in the headers, but when we add a provider that doesn't have a `kid` we can add support for this. 
```
func DiscoverPublicKey(ctx context.Context, headers jws.Headers, issuer string) (crypto.PublicKey, error) {
    discConf, err := oidcclient.Discover(issuer, http.DefaultClient)
    if err != nil {
        return nil, fmt.Errorf("failed to call OIDC discovery endpoint: %w", err)
    }


    jwks, err := jwk.Fetch(ctx, discConf.JwksURI)
    if err != nil {
        return nil, fmt.Errorf("failed to fetch to JWKS: %w", err)
    }


    kid := headers.KeyID()
    key, ok := jwks.LookupKeyID(kid)
    if !ok {
        return nil, fmt.Errorf("key %q isn't in JWKS", kid)
    }
```
By changing the PublicKey on an OP and Verifier to no longer accept headers we lose the flexibility to add feature without changing the OpenIdProvider and VerifierProvider interface.

Suggested Fix:  We've already talking about reverting the changes to the OpenIdProvider  interface so that should address this as well. I just want to leave a note here to remember to check that we didn't break the path to look up PublicKeys by jtk claims.